plugins/contents/fps_contents/fileid.py

```diff
@@ -0,0 +1,194 @@
+            self.initialized.set()
+
+            async for changes in awatch("."):
+                deleted_paths = []
+                added_paths = []
+                for change, changed_path in changes:
+                    # get relative path
+                    changed_path = Path(changed_path).relative_to(await Path().absolute())
+                    changed_path_str = str(changed_path)
+
+                    if change == Change.deleted:
+                        logger.debug("File %s was deleted", changed_path_str)
+                        async with db.execute(
+                            "SELECT COUNT(*) FROM fileids WHERE path = ?", (changed_path_str,)
+                        ) as cursor:
+                            if not (await cursor.fetchone())[0]:
+                                # path is not indexed, ignore
+                                logger.debug("File %s is not indexed, ignoring", changed_path_str)
+                                continue
+                        # path is indexed
+                        await maybe_rename(db, changed_path_str, deleted_paths, added_paths, False)
+                    elif change == Change.added:
+                        logger.debug("File %s was added", changed_path_str)
+                        await maybe_rename(db, changed_path_str, added_paths, deleted_paths, True)
+                    elif change == Change.modified:
+                        logger.debug("File %s was modified", changed_path_str)
+                        if changed_path_str == self.db_path:
+                            continue
+                        async with db.execute(
+                            "SELECT COUNT(*) FROM fileids WHERE path = ?", (changed_path_str,)
+                        ) as cursor:
+                            if not (await cursor.fetchone())[0]:
+                                # path is not indexed, ignore
+                                logger.debug("File %s is not indexed, ignoring", changed_path_str)
+                                continue
+                        mtime = (await changed_path.stat()).st_mtime
+                        await db.execute(
+                            "UPDATE fileids SET mtime = ? WHERE path = ?", (mtime, changed_path_str)
+                        )
+
+                for path in deleted_paths + added_paths:
+                    await db.execute("DELETE FROM fileids WHERE path = ?", (path,))
+                await db.commit()
+
+                for change in changes:
+                    changed_path = change[1]
+                    # get relative path
+                    changed_path = str(Path(changed_path).relative_to(await Path().absolute()))
+                    for watcher in self.watchers.get(changed_path, []):
+                        watcher.notify(change)
+
+    def watch(self, path: str) -> Watcher:
+        watcher = Watcher(path)
+        self.watchers[path] = self.watchers.get(path, [])
+        self.watchers[path].append(watcher)
```

Comment:
how about using dict.setdefault
```suggestion
        self.watchers.setdefault(path, []).append(watcher)
```

Reply:
Thanks @Wh1isper, I learned something!